src/core/include/units/quantity.h

```diff
@@ -392,25 +404,41 @@ template<Quantity Q1, QuantityEquivalentTo<Q1> Q2>
   return ret(ret(lhs).count() - ret(rhs).count());
 }
 
-template<typename D1, typename U1, typename Rep1, typename D2, typename U2, typename Rep2>
-  requires quantity_value_for_<std::multiplies<>, Rep1, Rep2>
-[[nodiscard]] constexpr Quantity auto operator*(const quantity<D1, U1, Rep1>& lhs, const quantity<D2, U2, Rep2>& rhs)
+template<typename QuantityOrQuantityValue, typename D, typename U>
+  requires Quantity<QuantityOrQuantityValue> || QuantityValue<QuantityOrQuantityValue>
+[[nodiscard]] constexpr Quantity auto operator*(const QuantityOrQuantityValue& lhs, reference<D, U>)
 {
-  using dim = dimension_multiply<D1, D2>;
-  using unit = downcast_unit<dim, (U1::ratio / dimension_unit<D1>::ratio) * (U2::ratio / dimension_unit<D2>::ratio) * dimension_unit<dim>::ratio>;
-  using ret = quantity<dim, unit, std::invoke_result_t<std::multiplies<>, Rep1, Rep2>>;
-  return ret(lhs.count() * rhs.count());
+  if constexpr (Quantity<QuantityOrQuantityValue>)
+    return lhs * quantity<D, U, typename QuantityOrQuantityValue::rep>::one();
+  else
+    return quantity<D, U, QuantityOrQuantityValue>(lhs);
 }
 
-template<typename D1, typename U1, typename Rep1, typename D2, typename U2, typename Rep2>
-  requires quantity_value_for_<std::divides<>, Rep1, Rep2>
-[[nodiscard]] constexpr Quantity auto operator/(const quantity<D1, U1, Rep1>& lhs, const quantity<D2, U2, Rep2>& rhs)
+template<Quantity Q1, Quantity Q2>
+  requires quantity_value_for_<std::multiplies<>, typename Q1::rep, typename Q2::rep>
+[[nodiscard]] constexpr Quantity auto operator*(const Q1& lhs, const Q2& rhs)
 {
-  gsl_ExpectsAudit(rhs.count() != quantity_values<Rep2>::zero());
-  using dim = dimension_divide<D1, D2>;
-  using unit = downcast_unit<dim, (U1::ratio / dimension_unit<D1>::ratio) / (U2::ratio / dimension_unit<D2>::ratio) * dimension_unit<dim>::ratio>;
-  using ret = quantity<dim, unit, std::invoke_result_t<std::divides<>, Rep1, Rep2>>;
-  return ret(lhs.count() / rhs.count());
+  return detail::make_quantity<reference_multiply<typename Q1::reference, typename Q2::reference>>(
+    lhs.count() * rhs.count());
+}
```

Comment:
Indeed. I just wanted to avoid instantiating references of unknown coherent units.

> Maybe it would be better to provide `operator*` and `operator/` for them?
> 
> Beside the above, this would also allow the use cases like:
> 
> ```c++
> inline constexpr auto km_per_h = km / h;
> ```
> 

That's already valid. I should refactor to make the interface of `reference` more visible. It's all burried under TMP and in `quantity.h` due to dependencies.

Reply:
>  I just wanted to avoid instantiating references of unknown coherent units.

I am not sure what is the difference here? Does making operators NOT hidden-friends non-member functions help?

Reply:
In the current state, I'm only using specializations of `reference` within `quantity`'s operators, along with (`divide`/`multiply`)`_reference`. So it never instantiates `reference` or its operators. They're only ever instantiated when defining unit constants or explicitly operating between references (like `km / s`). I do personally prefer the value approach, though.

Reply:
Ahh, by "instantiating" you mean creating an object of a type and not class template instantiations 😉

Reply:
I am not sure if I see the difference? What is the difference between instantiating `multiply_reference` class template and `operator*()` function template (besides the fact that the function template instantiation may be a bit slower at compile time). Either of them will be needed to be instantiated even if we will have an unknown coherent unit, right?

Reply:
Or did you mean the arguments and not the operation itself?

Reply:
(`divide`/`multiply`)`_reference` are template aliases, so they're [cheaper](https://www.reddit.com/r/cpp/comments/6gur2x/the_rule_of_chiel_aka_compiletime_cost_of/).